### PR TITLE
Fix/tool calls with chat endpoint

### PR DESCRIPTION
**Pull request by MotzWanted**
This PR introduces raffle-specific updates to `vllm`:

- Resolved a `TypeError` for pickle operations, to include json schemas in the request.
- Added functionality for structured text generation, enabling more precise output control.
- Introduced support for tool calls in the chat completion endpoint, expanding functionality.
- Optimized by removing extraneous logging for small speed-up.
- Added the `qwen1.5 14B` model in the `Dockerfile` for deployment.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -101,5 +101,5 @@ RUN --mount=type=cache,target=/root/.cache/pip \
 COPY --from=build /workspace/vllm/*.so /workspace/vllm/
 COPY vllm vllm
 
-ENTRYPOINT ["python3", "-m", "vllm.entrypoints.openai.api_server"]
+ENTRYPOINT ["python3", "-m", "vllm.entrypoints.openai.api_server","--model Qwen/Qwen1.5-14B-Chat"]
```
Comment (pdarulewski) [inline Dockerfile:104]: does it work? I'd say every parameter should be in a separate argument:
```suggestion
ENTRYPOINT ["python3", "-m", "vllm.entrypoints.openai.api_server","--model", "Qwen/Qwen1.5-14B-Chat"]
```

File: vllm/entrypoints/openai/serving_chat.py [hunk 2]
```diff
@@ -231,21 +261,77 @@ async def chat_completion_full_generator(
...
+        created_time = int(time.monotonic())
+
+        final_res: RequestOutput = None
+        async for request_output in result_generator:
+            if await raw_request.is_disconnected():
+                # Abort the request if the client disconnects.
+                await self.engine.abort(request_id)
+                return self.create_error_response("Client disconnected")
+            final_res = request_output
+        assert final_res is not None
```
Comment (pdarulewski) [inline vllm/entrypoints/openai/serving_chat.py:302]: let's avoid asserts:
- exceptions are better
- asserts should be only used in tests
- if you run the code with optimisation flags, e.g. O2, all asserts are ignored entirely

File: vllm/entrypoints/openai/serving_chat.py [hunk 3]
```diff
@@ -231,21 +261,77 @@ async def chat_completion_full_generator(
...
+            )
+
+        model_name = request.model
+        tool = request.tool
+        created_time = int(time.monotonic())
+
+        final_res: RequestOutput = None
+        async for request_output in result_generator:
+            if await raw_request.is_disconnected():
+                # Abort the request if the client disconnects.
```
Comment (pdarulewski) [inline vllm/entrypoints/openai/serving_chat.py:298]: this comment is not really needed, you can figure out stuff based on the if condition 👍

File: vllm/entrypoints/openai/logit_processors.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        """
+        tokenizer.vocabulary = tokenizer.get_vocab()
+        tokenizer.special_tokens = set(tokenizer.all_special_tokens)
+
+        def convert_token_to_string(token: str) -> str:
+            from transformers.file_utils import SPIECE_UNDERLINE
+
+            string = tokenizer.convert_tokens_to_string([token])
+
+            # A hack to handle missing spaces to HF's Llama tokenizers
```
Comment (pdarulewski) [inline vllm/entrypoints/openai/logit_processors.py:68]: can you explain this "hack"? Or provide a link to GitHub / StackOverflow issue?

File: vllm/engine/async_llm_engine.py [hunk 5]
```diff
@@ -426,16 +446,15 @@ async def add_request(
...
-                    shortened_token_ids = shortened_token_ids[:self.
-                                                              max_log_len]
-            logger.info(f"Received request {request_id}: "
-                        f"prompt: {shortened_prompt!r}, "
-                        f"prefix_pos: {prefix_pos},"
-                        f"sampling_params: {sampling_params}, "
-                        f"prompt_token_ids: {shortened_token_ids}, "
-                        f"lora_request: {lora_request}.")
+                    shortened_token_ids = shortened_token_ids[: self.max_log_len]
+            # logger.info(f"Received request {request_id}: "
```
Comment (pdarulewski) [inline vllm/engine/async_llm_engine.py:452]: no logging here?

Comment (MotzWanted) [reply vllm/engine/async_llm_engine.py:452]: It's a quite extensive log slowing requests down. We're already logging the http status code when receiving a request, the throughput every 5 second or something, and we have this endpoint to monitor requests with grafana: https://github.com/raffle-ai/vllm/blob/c530e2cfe3b3d7e60130ff817cee7f3a395af232/vllm/entrypoints/openai/api_server.py#L145

Comment (pdarulewski) [reply vllm/engine/async_llm_engine.py:452]: I knew it's gonna happen 😁
Nice! How do they monitor the requests? Do they use prometheus?

**Review by pdarulewski — APPROVED** [review]

**Pull request merged**